azure-devops/create-1es-hosted-pool.ps1

```diff
@@ -316,130 +288,159 @@ Invoke-AzVMRunCommand `
-Write-Progress `
-  -Activity $ProgressActivity `
-  -Status 'Deleting unused VM and disk' `
-  -PercentComplete (100 / $TotalProgress * $CurrentProgress++)
+Display-Progress-Bar -Status 'Creating gallery'
 
-Remove-AzVM -Id $VM.ID -Force | Out-Null
-Remove-AzDisk `
+$GalleryName = 'StlBuild_' + $CurrentDate.ToString('yyyy_MM_dd_THHmm') + '_Gallery'
+$Gallery = New-AzGallery `
+  -Location $Location `
   -ResourceGroupName $ResourceGroupName `
-  -DiskName $PrototypeOSDiskName `
-  -Force | Out-Null
+  -Name $GalleryName
+
+####################################################################################################
+Display-Progress-Bar -Status 'Granting access to 1ES Resource Management'
+
+$ServicePrincipalObjectId = (Get-AzADServicePrincipal -DisplayName '1ES Resource Management').Id
+
+New-AzRoleAssignment `
+  -ObjectId $ServicePrincipalObjectId `
+  -RoleDefinitionName 'Reader' `
+  -Scope $Gallery.Id | Out-Null
 
 ####################################################################################################
-Write-Progress `
-  -Activity $ProgressActivity `
-  -Status 'Creating scale set' `
-  -PercentComplete (100 / $TotalProgress * $CurrentProgress++)
-
-$VmssIpConfigName = $ResourceGroupName + '-VmssIpConfig'
-$VmssIpConfig = New-AzVmssIpConfig -SubnetId $Nic.IpConfigurations[0].Subnet.Id -Primary -Name $VmssIpConfigName
-$VmssName = $ResourceGroupName + '-Vmss'
-$Vmss = New-AzVmssConfig `
+Display-Progress-Bar -Status 'Creating image definition'
+
+$ImageDefinitionName = $ResourceGroupName + '-ImageDefinition'
+New-AzGalleryImageDefinition `
   -Location $Location `
-  -SkuCapacity 0 `
-  -SkuName $VMSize `
-  -SkuTier 'Standard' `
-  -Overprovision $false `
-  -UpgradePolicyMode Manual `
-  -EvictionPolicy Delete `
-  -Priority Spot `
-  -MaxPrice -1
+  -ResourceGroupName $ResourceGroupName `
+  -GalleryName $GalleryName `
+  -Name $ImageDefinitionName `
+  -OsState 'Generalized' `
+  -OsType 'Windows' `
+  -Publisher $ImagePublisher `
+  -Offer $ImageOffer `
+  -Sku $ImageSku `
+  -HyperVGeneration 'V2' | Out-Null
 
-$Vmss = Add-AzVmssNetworkInterfaceConfiguration `
-  -VirtualMachineScaleSet $Vmss `
-  -Primary $true `
-  -IpConfiguration $VmssIpConfig `
-  -NetworkSecurityGroupId $NetworkSecurityGroup.Id `
-  -Name $NicName
-
-$Vmss = Set-AzVmssOsProfile `
-  -VirtualMachineScaleSet $Vmss `
-  -ComputerNamePrefix $LiveVMPrefix `
-  -AdminUsername 'AdminUser' `
-  -AdminPassword $AdminPW `
-  -WindowsConfigurationProvisionVMAgent $true `
-  -WindowsConfigurationEnableAutomaticUpdate $true
-
-$Vmss = Set-AzVmssStorageProfile `
-  -VirtualMachineScaleSet $Vmss `
-  -OsDiskCreateOption 'FromImage' `
-  -OsDiskCaching ReadWrite `
-  -ImageReferenceId $Image.Id
-
-$Vmss = New-AzVmss `
+####################################################################################################
+Display-Progress-Bar -Status 'Creating image version'
+
+$ImageVersionName = $CurrentDate.ToString('yyyyMMdd.HHmm.0')

```

Comment:
Is the `.0` required here, or could we omit it? It seems unlikely that we'd ever run this script more than once a minute.

Reply:
It is required: https://docs.microsoft.com/en-us/azure/virtual-machines/image-version?tabs=powershell

> Allowed characters for the image version are numbers and periods. Numbers must be within the range of a 32-bit integer. Format: *MajorVersion.MinorVersion.Patch*.

We could use `1.0.0` since the compute gallery and image definition provide uniqueness, but I thought that using the date and time was a good idea to be consistent (now they all use the date and time), be globally unique (during periods when we have old and new pools), and be possibly future-proofed (if we ever need to change to a scheme where we have multiple image versions).

Reply:
Yes, the date and time is certainly preferable. Just double-checking. (This comment is implicitly "no change requested", since I approved.)